### PR TITLE
Helper function to enable bucketing

### PR DESCRIPTION
**Pull request by ruisizhang123**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #19
* #18
* #17
* __->__ #16

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -4461,9 +4457,12 @@ def process_kernel(
             else:
                 if isinstance(arg, sympy.Expr):
                     arg = V.graph.sizevars.shape_env.create_symintnode(arg, hint=None)
+                # if isinstance(arg, FakeTensor):
+                #    tensor_args.append(arg)
```
Comment (tianyu-l) [inline torch/_inductor/ir.py]: pls remove

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -4461,9 +4457,12 @@ def process_kernel(
...
                 non_tensor_args.append(arg)
 
         def unflatten_args(new_tensor_args, new_non_tensor_args):
+
```
Comment (tianyu-l) [inline torch/_inductor/ir.py]: remove empty line

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -4475,8 +4474,13 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):
...
 
-        tensor_args = [cls.realize_input(x) for x in tensor_args]
-
+        tensor_args_new = []
+        for x in tensor_args:
+            if process_kernel_seperate:
+                tensor_args_new.append(TensorBox(StorageBox(x)))
+            else:
+                tensor_args_new.append(cls.realize_input(x))
+        tensor_args = tensor_args_new
```
Comment (tianyu-l) [inline torch/_inductor/ir.py]:
```suggestion
        if process_kernel_seperate:
            tensor_args = [TensorBox(StorageBox(x)) for x in tensor_args]
        else:
            tensor_args = [cls.realize_input(x) for x in tensor_args]
```

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -4492,27 +4496,23 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):
...
-            ):
+            elif x.get_name() in V.graph.torchbind_constants:
                 example_args.append(V.graph.torchbind_constants[x.get_name()])
             else:
                 example_args.append(ir_node_to_tensor(x, guard_shape=True))
 
         new_args, new_kwargs = unflatten_args(example_args, non_tensor_args)
-        example_output = kernel(*new_args, **new_kwargs)
 
+        example_output = kernel(*new_args, **new_kwargs)
```
Comment (tianyu-l) [inline torch/_inductor/ir.py:4502]: put this back to where it was?

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -4492,27 +4496,23 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):
         # propagated the graph with, because for some operators running without a
         # constant would trigger an error / DataDependentException
         for x in tensor_args:
-            # if x is a view of a constant, we need to realize the view
-            # (we can't pass the constant into the kernel directly)
-            if not isinstance(x, BaseView) and x.get_name() in V.graph.constants:
+            if x.get_name() in V.graph.constants:
```
Comment (tianyu-l) [inline torch/_inductor/ir.py:4493]: is this modification needed?

File: torch/_inductor/ir.py [hunk 6]
```diff
@@ -4492,27 +4496,23 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):
...
                 example_args.append(V.graph.constants[x.get_name()])
-            elif (
-                not isinstance(x, BaseView)
-                and x.get_name() in V.graph.torchbind_constants
-            ):
+            elif x.get_name() in V.graph.torchbind_constants:
```
Comment (tianyu-l) [inline torch/_inductor/ir.py:4495]: ditto

File: torch/_inductor/ir.py [hunk 7]
```diff
@@ -4492,27 +4496,23 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):
...
-        example_output = kernel(*new_args, **new_kwargs)
 
+        example_output = kernel(*new_args, **new_kwargs)
         unbacked_bindings: Optional[Dict[sympy.Symbol, pytree.KeyPath]] = None
-        if shape_env := V.fake_mode.shape_env:
-            rebind_unbacked(shape_env, V.current_node, example_output)
-            unbacked_bindings = compute_unbacked_bindings(
-                shape_env, example_output, V.current_node.meta.get("val")
-            )
+        if not isinstance(V.current_node, NullHandler):
```
Comment (tianyu-l) [inline torch/_inductor/ir.py:4504]: it'd be great if you could add comments, e.g. what happens when `isinstance(V.current_node, NullHandler)` and why that's needed for the following works.

File: torch/_inductor/ir.py [hunk 8]
```diff
@@ -4522,8 +4522,11 @@ def unflatten_args(new_tensor_args, new_non_tensor_args):
         for t in example_out_li:
             if isinstance(t, torch.Tensor) and t.is_sparse:
                 msg = "sparsity not handled. Please file issue for sparse inference weights."
-                if stack_trace := V.graph.current_node.meta.get("stack_trace", None):
-                    msg = f"{msg} Found from : \n {stack_trace}"
+                if not isinstance(V.current_node, NullHandler):
```
Comment (tianyu-l) [inline torch/_inductor/ir.py:4519]: same here

File: torch/_inductor/ir.py [hunk 9]
```diff
@@ -5949,6 +5956,15 @@ def codegen(self, wrapper):
             )
         else:
             self.codegen_comment(wrapper)
+            if (
+                self.op_overload == torch.ops.fsdp.split_with_sizes_copy.default
+                or self.op_overload == torch.ops.fsdp.read_out.default
+            ):
+                for buf in self.inputs[1:]:
+                    V.graph.wrapper_code.codegen_allocation(buf, force_alloc=True)
+            if self.op_overload == torch.ops.fsdp.chunk_cat.default:
```
Comment (tianyu-l) [inline torch/_inductor/ir.py]: why AG copy-in is missing? I wonder if it causes the memory regression bug.

File: torch/_inductor/ir.py [hunk 10]
```diff
@@ -7044,14 +7079,36 @@ def create_out_of_place(
...
+                isinstance(inputs, MultiOutput)
+                and isinstance(inputs.inputs[0], FallbackKernel)
+                and inputs.inputs[0].op_overload
+                in [
+                    torch.ops.fsdp.all_gather_copy_in.default,
+                    torch.ops.fsdp.chunk_cat.default,
+                ]
+            ) or (
+                isinstance(inputs, FallbackKernel)
+                and inputs.op_overload == torch.ops.fsdp.chunk_cat.default
```
Comment (tianyu-l) [inline torch/_inductor/ir.py]: ditto: why AG copy-in is missing

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 11]
```diff
@@ -100,14 +107,67 @@ def split_with_sizes_copy(
...
+        all_gather_input_split_sizes = [
+            a // world_size for a in all_gather_input_split_sizes
+        ]
+        torch.split_with_sizes_copy(
+            all_gather_output, all_gather_input_split_sizes, dim=dim, out=out
+        )
+    else:
+        torch.split_with_sizes_copy(
+            all_gather_output, all_gather_input_split_sizes, dim=dim, out=out
+        )
```
Comment (tianyu-l) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py]: we can reuse the code here

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 12]
```diff
@@ -118,9 +178,26 @@ def chunk_cat(
...
+        device = tensors[0].device
+        reduce_dtype = torch.bfloat16
+        padded_unsharded_sizes = tuple(
+            _get_dim0_padded_size(grad.size(), world_size) for grad in tensors
+        )
+        reduce_scatter_input_numel = sum(s.numel() for s in padded_unsharded_sizes)
+        chunk_cat_out = torch.empty(
+            (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
+        )
+        chunk_cat_out = chunk_cat_out.view(world_size, -1)
```
Comment (tianyu-l) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py]: why this variable is not used anymore?

**Review by tianyu-l — CHANGES_REQUESTED** [review]
looks much better! had some questions and suggestions

File: torch/_inductor/ir.py [hunk 13]
```diff
@@ -5972,13 +5988,26 @@ def create(cls, kernel, *args, **kwargs):
...
         )
         with context:
+
+            if (
+                kernel == torch.ops.fsdp.all_gather_copy_in.default
+                or kernel == torch.ops.fsdp.split_with_sizes_copy.default
+                or kernel == torch.ops.fsdp.chunk_cat.default
+                or kernel == torch.ops.fsdp.read_out.default
+            ):
+                process_kernel_seperate = True
```
Comment (tianyu-l) [inline torch/_inductor/ir.py]: in general I wonder what if these ops are from FSDP2 rather than Simple FSDP -- what would it cause?

**Review by tianyu-l — APPROVED** [review]
please
1. comment in the PR where I raised questions, in particular the changes you inherited from Will -- can you link to the PR he had?
2. `process_kernel_separate` only when `simplefsdp` is applied. Not sure what's the best way to do this, e.g. we can do it when `torch._inductor.simplefsdp.enable_bucket = True` (but you may need to move the definition to this PR).

**Comment by ruisizhang123** [comment]
> please
> 
> 1. comment in the PR where I raised questions, in particular the changes you inherited from Will -- can you link to the PR he had?

The diffs is https://gist.github.com/yf225/e2b2d424f190050dda92dde1ef63e408.

> 2. `process_kernel_separate` only when `simplefsdp` is applied. Not sure what's the best way to do this, e.g. we can do it when `torch._inductor.simplefsdp.enable_bucket = True` (but you may need to move the definition to this PR).

Moving the config to this pr seems to be the easiest way to me. I just updated

**Pull request merged**